unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/VoiceChatHUD/VoiceChatWindowController.cs

```diff
@@ -0,0 +1,436 @@
+    {
+        if (trackedUsersHashSet.Contains(userId))
+            trackedUsersHashSet.Remove(userId);
+
+        if (!currentPlayers.TryGetValue(userId, out VoiceChatPlayerComponentView elementView))
+            return;
+
+        if (!elementView)
+            return;
+
+        playersPool.Enqueue(elementView);
+        currentPlayers.Remove(userId);
+        voiceChatWindowView.SetNumberOfPlayers(currentPlayers.Count);
+
+        elementView.SetActive(false);
+
+        SetWhichPlayerIsTalking();
+        CheckMuteAllState();
+    }
+
+    internal void OnMuteAllToggled(bool isMute)
+    {
+        isMuteAll = isMute;
+
+        if (!isJoined)
+            return;
+
+        MuteAll(isMute);
+    }
+
+    internal void MuteAll(bool isMute)
+    {
+        isMuteAll = isMute;
+
+        if (isMute)
+            usersToUnmute.Clear();
+        else
+            usersToMute.Clear();
+
+        using (var iterator = trackedUsersHashSet.GetEnumerator())
+        {
+            while (iterator.MoveNext())
+            {
+                MuteUser(iterator.Current, isMute);
+            }
+        }
+    }
+
+    internal void MuteUser(string userId, bool isMuted)
+    {
+        var list = isMuted ? usersToMute : usersToUnmute;
+        list.Add(userId);
+
+        if (updateMuteStatusRoutine == null)
+            updateMuteStatusRoutine = CoroutineStarter.Start(MuteStateUpdateRoutine());
+
+        if (isMuted)
+            socialAnalytics.SendPlayerMuted(userId);
+        else
+            socialAnalytics.SendPlayerUnmuted(userId);
+    }
+
+    internal IEnumerator MuteStateUpdateRoutine()
+    {
+        yield return WaitForSecondsCache.Get(MUTE_STATUS_UPDATE_INTERVAL);
+        ReportMuteStatuses();
+        updateMuteStatusRoutine = null;
+    }
+
+    internal void ReportMuteStatuses()
+    {
+        if (usersToUnmute.Count > 0)
+            WebInterface.SetMuteUsers(usersToUnmute.ToArray(), false);
+
+        if (usersToMute.Count > 0)
+            WebInterface.SetMuteUsers(usersToMute.ToArray(), true);
+
+        usersToUnmute.Clear();
+        usersToMute.Clear();
+    }
+
+    internal void ChangeAllowUsersFilter(string optionId)
+    {
+        var newSettings = settings.generalSettings.Data;
+
+        if (optionId == VoiceChatAllow.ALL_USERS.ToString())
+            newSettings.voiceChatAllow = VoiceChatAllow.ALL_USERS;
+        else if (optionId == VoiceChatAllow.VERIFIED_ONLY.ToString())
+            newSettings.voiceChatAllow = VoiceChatAllow.VERIFIED_ONLY;
+        else if (optionId == VoiceChatAllow.FRIENDS_ONLY.ToString())
+            newSettings.voiceChatAllow = VoiceChatAllow.FRIENDS_ONLY;
+
+        settings.generalSettings.Apply(newSettings);
+    }
+
+    internal void OpenContextMenu(string userId)
+    {
+        currentPlayers.TryGetValue(userId, out VoiceChatPlayerComponentView elementView);
+        if (elementView != null)
+            elementView.DockAndOpenUserContextMenu(voiceChatWindowView.ContextMenuPanel);
```

Comment:
it should not be necessary to pass the contextual menu from the controller if the view handlers the internal views

Reply:
I'm following this way because the contextual menu is handled by the `voiceChatWindowView` but in this case we want to give this reference to an `elementView` that is another DIFFERENT view, so I think the correct place to make this relationship is the main controller of the feature that is already managing the several views.